src/clients/UBAClient/UBAClientBase.ts

```diff
@@ -0,0 +1,346 @@
+      .filter(
+        (flow) =>
+          (fromBlock === undefined || flow.blockNumber >= fromBlock) &&
+          (toBlock === undefined || flow.blockNumber <= toBlock)
+      );
+  }
+
+  /**
+   * @description Evaluate an RefundRequest object for validity.
+   * @dev  Callers should evaluate 'valid' before 'reason' in the return object.
+   * @dev  The following RefundRequest attributes are not evaluated for validity and should be checked separately:
+   * @dev  - previousIdenticalRequests
+   * @dev  - Age of blockNumber (i.e. according to SpokePool finality)
+   * @param chainId       ChainId of SpokePool where refundRequest originated.
+   * @param refundRequest RefundRequest object to be evaluated for validity.
+   */
+  public refundRequestIsValid(chainId: number, refundRequest: RefundRequestWithBlock): RequestValidReturnType {
+    const result = this.getFlows(chainId, refundRequest.refundToken).some((flow) => {
+      return (
+        flow.logIndex == refundRequest.logIndex &&
+        flow.blockNumber == refundRequest.blockNumber &&
+        flow.transactionHash == refundRequest.transactionHash &&
+        flow.transactionIndex == refundRequest.transactionIndex
+      );
+    });
+    if (!result) {
+      return {
+        valid: false,
+        reason: "RefundRequest is not a valid flow because it didn't appear in the list of validated flows.",
+      };
+    } else {
+      return {
+        valid: true,
+      };
+    }
+  }
+
+  /**
+   * Calculate the balancing fee of a given token on a given chainId at a given block number
+   * @param tokenSymbol The token to get the balancing fee for
+   * @param amount The amount to get the balancing fee for
+   * @param balancingActionBlockNumber The block number to get the balancing fee for
+   * @param chainId The chainId to get the balancing fee for. If the feeType is Deposit, this is the deposit chainId. If the feeType is Refund, this is the refund chainId.
+   * @param feeType The type of fee to calculate
+   * @returns The balancing fee for the given token on the given chainId at the given block number
+   */
+  public computeBalancingFee(
+    tokenSymbol: string,
+    amount: BigNumber,
+    balancingActionBlockNumber: number,
+    chainId: number,
+    feeType: UBAActionType
+  ): BalancingFeeReturnType {
+    // Opening balance for the balancing action blockNumber.
+    const relevantBundleStates = this.retrieveBundleStates(chainId, tokenSymbol);
+    const specificBundleState = relevantBundleStates.findLast(
+      (bundleState) => bundleState.blockNumber <= balancingActionBlockNumber
+    );
+    if (!specificBundleState) {
+      throw new Error(`No bundle states found for token ${tokenSymbol} on chain ${chainId}`);
+    }
+    /** @TODO ADD TX INDEX COMPARISON */
+    const flows = (specificBundleState?.flows ?? []).filter(
+      (flow) => flow.flow.blockNumber <= balancingActionBlockNumber
+    );
+    const calculator = new UBAFeeSpokeCalculator(
```

Comment:
Just a suggestion but isn't it kind of strange to instantiate a new calculator each time we call `computeBalancingFee`? Should we instead treat the calculator as a library with a bunch of static functions and make calls like:

```ts
const { balancingFee } = UBAFeeSpokeCalculator[feeType === UBAActionType.Deposit ? "getDepositFee" : "getRefundFee"](
    chainId, 
    tokenSymbol, 
    flows.map(({ flow }) => flow), 
    specificBundleState.openingBalance,
    specificBundleState.openingIncentiveBalance,
    specificBundleState.config.ubaConfig,
    amount
)
```

Reply:
Sounds good - I've created analog versions of all the UBA Spoke Fee calculators so you can run it without instantiating a class: https://github.com/across-protocol/sdk-v2/pull/266/commits/6cd922196f94c49119642da9940dc501814c0ece

Reply:
I don't see us using the analog version in This function. Shouldn't we?